src/libpmemfile-posix/file.c

```diff
@@ -0,0 +1,2585 @@
+		return EOPNOTSUPP;
+	}
+
+	if (mode & PMEMFILE_FL_PUNCH_HOLE) {
+		/*
+		 * from man 2 fallocate:
+		 *
+		 * "The FALLOC_FL_PUNCH_HOLE flag must be ORed
+		 * with FALLOC_FL_KEEP_SIZE in mode; in other words,
+		 * even when punching off the end of the file, the file size
+		 * (as reported by stat(2)) does not change."
+		 */
+		if (mode != (PMEMFILE_FL_PUNCH_HOLE | PMEMFILE_FL_KEEP_SIZE))
+			return EINVAL;
+	} else { /* Allocating disk space */
+		/*
+		 * Note: According to 'man 2 fallocate' FALLOC_FL_UNSHARE
+		 * is another possible flag to accept here. No equivalent of
+		 * that flag is supported by pmemfile as of now. Also that man
+		 * page is wrong anyways, the header files only refer to
+		 * FALLOC_FL_UNSHARE_RANGE, so it is suspected that noone is
+		 * using it anyways.
+		 */
+		if ((mode & ~PMEMFILE_FL_KEEP_SIZE) != 0)
+			return EINVAL;
+	}
+
+	return 0;
+}
+
+/*
+ * file_fallocate - part of pmemfile_fallocate implementation
+ * Operates on a PMEMfile, and calls vinode_fallocate, which
+ * does the actual fallocate operation on the inode.
+ *
+ * This function expected the file to locked while being called,
+ * and makes sure the vinode is locked while calling vinode_fallocate.
+ */
+static int
+file_fallocate(PMEMfilepool *pfp, PMEMfile *file, int mode,
+		uint64_t offset, uint64_t length)
+{
+	int error;
+
+	/*
+	 * from man 2 fallocate:
+	 *
+	 * "EBADF  fd is not a valid file descriptor, or is not opened for
+	 * writing."
+	 */
+	if ((file->flags & PFILE_WRITE) == 0)
+		return EBADF;
+
+	os_rwlock_wrlock(&file->vinode->rwlock);
+
+	error = vinode_fallocate(pfp, file->vinode, mode, offset, length);
+
+	os_rwlock_unlock(&file->vinode->rwlock);
+
+	return error;
+}
+
+int
+pmemfile_fallocate(PMEMfilepool *pfp, PMEMfile *file, int mode,
+		pmemfile_off_t offset, pmemfile_off_t length)
+{
+	if (!pfp) {
+		LOG(LUSR, "NULL pool");
+		errno = EFAULT;
+		return -1;
+	}
+
+	if (!file) {
+		LOG(LUSR, "NULL file");
+		errno = EFAULT;
+		return -1;
+	}
+
+	int error;
+
+	error = fallocate_check_arguments(mode, offset, length);
+
+	if (error == 0) {
+		ASSERT(offset >= 0);
+		ASSERT(length > 0);
+
+		os_mutex_lock(&file->mutex);
+
+		error = file_fallocate(pfp, file, mode,
+		    (uint64_t)offset, (uint64_t)length);
+
+		os_mutex_unlock(&file->mutex);
+	}
+
+	if (error != 0) {
+		errno = error;
+		return -1;
+	}
+
+	return 0;
+}
+
+int
+pmemfile_posix_fallocate(PMEMfilepool *pfp, PMEMfile *file,
+		pmemfile_off_t offset, pmemfile_off_t length)
+{
+	return pmemfile_fallocate(pfp, file, 0, offset, length);
+}
+
+/*
+ * vinode_chown
+ *
+ * Can't be called in a transaction.
+ */
+static int
+vinode_chown(PMEMfilepool *pfp, const struct pmemfile_cred *cred,
+		struct pmemfile_vinode *vinode, pmemfile_uid_t owner,
+		pmemfile_gid_t group)
+{
+	struct pmemfile_inode *inode = vinode->inode;
+	int error = 0;
+
+	ASSERTeq(pmemobj_tx_stage(), TX_STAGE_NONE);
+
+	if (owner == (pmemfile_uid_t)-1 && group == (pmemfile_gid_t)-1)
```

Comment:
It isn't. Chown mapage says:
"If the owner or group is specified as -1, then that ID is not changed."